src/main/java/com/levik/pool/CustomConnectionDataSource.java

```diff
@@ -0,0 +1,96 @@
+package com.levik.pool;
+
+import lombok.SneakyThrows;
+import org.postgresql.ds.PGSimpleDataSource;
+
+import javax.sql.DataSource;
+import java.io.PrintWriter;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.logging.Logger;
+
+public class CustomConnectionDataSource implements DataSource {
+
+    private static final int DEFAULT_CONNECTION_POOL_SIZE = 10;
+    private final Queue<ConnectionWrapper> connectionPool = new ArrayDeque<>();
+
+    private final String url;
+    private final String username;
+    private final String password;
```

Comment:
looks like all these variables are used only once when we create an object, let's send it as parameters to the method, or prepare some object and also send as a parameter which will be deleted by GC after object creation